README.md
<p align="center">

<img width="162" height="100" src="https://raw.githubusercontent.com/MarwanAlsoltany/velox/master/themes/velox/assets/images/velox-logo.png" alt="VELOX"/>

</p>


---


<h1 align="center">VELOX</h1>

<div align="center">

The fastest way to build simple websites using PHP!


[![PHP Version][php-icon]][php-href]
[![Latest Version on Packagist][version-icon]][version-href]
[![Total Downloads][downloads-icon]][downloads-href]
[![License][license-icon]][license-href]
[![Maintenance][maintenance-icon]][maintenance-href]
[![Documentation][documentation-icon]][documentation-href]
[![Scrutinizer Build Status][scrutinizer-icon]][scrutinizer-href]
[![Scrutinizer Code Coverage][scrutinizer-coverage-icon]][scrutinizer-coverage-href]
[![Scrutinizer Code Quality][scrutinizer-quality-icon]][scrutinizer-quality-href]
[![Travis Build Status][travis-icon]][travis-href]
[![StyleCI Code Style][styleci-icon]][styleci-href]

[![Open in Visual Studio Code][vscode-icon]][vscode-href] [![Run on Repl.it][replit-icon]][replit-href]

[![Tweet][tweet-icon]][tweet-href] [![Star][github-icon]][github-href]



<details>
<summary>Table of Contents</summary>
<p>

[Installation](#installation)<br/>
[About VELOX](#about-velox)<br/>
[Architecture](#architecture)<br/>
[Config](#config)<br/>
[Classes](#classes)<br/>
[Functions](#functions)<br/>
[Commands](#commands)<br/>
[Themes](#themes)<br/>
[Extending VELOX](#extending-velox)<br/>
[MVC](#mvc)<br/>
[Templating](#templating)<br/>
[Authentication](#authentication)<br/>
[Changelog](./CHANGELOG.md)<br/>
[Documentation](https://marwanalsoltany.com/velox)

</p>
</details>

<br/>

<sup>If you like this project and would like to support its development, giving it a :star: would be appreciated!</sup>

<br/>

<a href="https://velox.marwanalsoltany.repl.co">
<img src="https://user-images.githubusercontent.com/7969982/133680020-48b64343-eb6e-4519-b1ba-adc86737c184.jpeg" alt="VELOX Demo" width="768" />
</a>

Check out the [**Demo**](https://velox.marwanalsoltany.repl.co) or play with the [**REPL**](https://replit.com/@MarwanAlsoltany/velox).

Do you feel like reading? Check out the full API on the documentation website on [`marwanalsoltany.github.io/velox`](https://marwanalsoltany.github.io/velox).

</div>


---


## Key Features

1. Zero dependencies
2. Minimal, intuitive and easy to get along with
3. Unlimited flexibility when it comes to customizing it to your exact needs

---


## Installation

#### Using Composer:

```sh
composer create-project marwanalsoltany/velox my-velox-app
```

![RED](https://fakeimg.pl/11/f03c15/000000?text=+) **Note:** *You may need to add the `--stability=dev` depending on the version/branch. You may also want to add `--no-dev` flag to not install development dependencies.*

#### Using Git:

```sh
git clone https://github.com/MarwanAlsoltany/velox.git my-velox-app
```

![BLUE](https://fakeimg.pl/11/1e90ff/000000?text=+) **Fact:** *If you don't want to use any other third party packages. Installing VELOX using Git is sufficient.*

#### Using Source:

Download [VELOX](https://github.com/MarwanAlsoltany/velox/releases) as a `.zip` or `.tar.gz` and extract it in your server web root directory.

![GREEN](https://fakeimg.pl/11/32cd32/000000?text=+) **Advice:** *If you want to test out VELOX quickly and you don't have any web server available, use whatever installing method and run `php bin/app-serve` from inside VELOX directory. This command will spin up a development web server on `localhost:8000` (Note that you need to have at least PHP installed on your system).*


---


## About VELOX

VELOX is a lightweight micro-framework that makes creating a simple website using PHP joyful. It helps you create future-proof websites faster and more efficiently. It provides components that facilitate the process of creating a website using PHP. VELOX does not have any dependencies, the VELOX package and everything that it needs is included in the project itself. All that VELOX provides is a way to work with **config**, pass **data**, register **routes**, interact with the **database**, render **views**, handle **exceptions**, **autoload** code, and **resolve** assets. It provides the *View* and the *Controller* parts of an *MVC* design pattern. Staring from `v1.3.0`, VELOX also provides the *Model* part, making it a fully featured *MVC* framework and starting from `v1.4.0` it also comes shipped with a simple authentication system. VELOX can also be used as a **Static Site Generator** if all you need is HTML files in the end.

### Why does VELOX exist?

VELOX was created to solve a specific problem, it's a way to build a website that is between dynamic and static, a way to create a simple website with few pages without being forced to use a framework or a CMS that comes with a ton of stuff which will never get used, it's lightweight, minimal, and straight to the point.

It's not recommended to use VELOX if you have an intermediary project, you would be better off using a well-established framework. VELOX is not an initiative to reinvent the wheel, you can look at VELOX as a starter-kit for small projects.

VELOX has a very special use-case, simple websites, and here is meant really simple websites. The advantage is, you don't have stuff that you don't need. Comparing VELOX to Laravel or Symfony is irrelevant, as these frameworks play in a totally different area, it also worth mentioning that VELOX is much simpler than Lumen or Slim.


---


## Architecture

### Directory structure

| Directory | Description |
| --- | --- |
| [`bootstrap`](./bootstrap) | This is where VELOX bootstraps the application. You normally don't have to change anything in this directory, unless you want to extend VELOX functionality beyond basic stuff. |
| [`bin`](./bin) | This is where PHP executables are placed. You can freely add yours, or delete the entire directory. |
| [`app`](./app) | This is where your own backend logic will be placed. You will be mostly working here for the backend part of your app. |
| [`classes`](./classes) | This is where VELOX source files live. You shouldn't be touching anything here unless you want to make your own version of VELOX. |
| [`functions`](./functions) | This is where all functions that are loaded in the application live. You can freely add yours, or delete the entire directory.|
| [`includes`](./includes) | This is where all files that should be preloaded will be placed. You can freely add yours, or delete the entire directory.|
| [`themes`](./themes) | This is where all your frontend themes will be placed. You will be mostly working here for the frontend part of your app. |
| [`config`](./config) | This is where all config files will live. All files here will be accessible using the `Config` class at runtime. |
| [`storage`](./storage) | This is where VELOX will write caches and logs. You can also use this directory to store installation-wide assets. |
| [`public`](./public) | This is where you should put your `index.php` with a symlink for static assets (active theme `assets/` directory for example) for maximum security. You can freely delete this directory if you want to. |
| [`vendor`](./vendor) | This is where your Composer dependencies will be placed. You can freely delete this directory if you don't want to use Composer. |

![GREEN](https://fakeimg.pl/11/32cd32/000000?text=+) **Advice:** *Most files listed in these directories are documented. Take a look through them to learn more about VELOX.*

### App Entry

The entry point for a VELOX app is the [`index.php`](./index.php), here you need to require the [`bootstrap/autoload.php`](./bootstrap/autoload.php), register some routes with their handlers using the `Router::class`, and start the router. This is all that you need to have a working VELOX app.

```php
<?php

require 'bootstrap/autoload.php';


Router::handle('/', function () {
    return View::render('home', ['title' => 'Home']);
});

Router::start();
```

Additionally, you can add middlewares using `Router::middleware()` and/or set up handlers for `404` and `405` responses using `{global.errorPages.CODE}` config value.

Alternatively, you can extract the *"routes registration part"* in its own file and let VELOX know about it using [`bootstrap/additional.php`](./bootstrap/additional.php).
Starting from `v1.2.0` VELOX does that by default, the file [`includes/routes/web.php`](./includes/routes/web.php) is where you should register your routes. The router will also start automatically if not started explicitly.

![RED](https://fakeimg.pl/11/f03c15/000000?text=+) **Note:** *In order for VELOX to work correctly and safely, you need to redirect all requests to application entry point (`index.php`) and block all requests to other PHP files on the server (take a look at [`.htaccess.dist`](./.htaccess.dist) to get started with Apache).*


---


## Config

The following table lists all config files that come shipped with VELOX.

| Config File | Description |
| --- | --- |
| [`global.php`](./config/global.php) | This config file contains some global variables that are used by almost all classes (app-wide config). |
| [`router.php`](./config/router.php) | This config file can be used to override `Router::class` default parameters. |
| [`session.php`](./config/session.php) | This config file contains session configuration, it is used by the `Session::class`. |
| [`database.php`](./config/database.php) | This config file contains database credentials, it is used by the `Database::class`. |
| [`auth.php`](./config/auth.php) | This config file contains authentication configuration, it is used by the `Auth::class`. |
| [`theme.php`](./config/theme.php) | This config file can be used to edit/extend theme configuration. |
| [`view.php`](./config/view.php) | This config file can be used to customize everything about the views. It is used by the `View::class`. |
| [`data.php`](./config/data.php) | This config file can be used to provide any arbitrary data, which then will get injected in the `Data::class`. |
| [`cli.php`](./config/cli.php) | This config file can be used to enable/disable the commands or change their arguments. |

![BLUE](https://fakeimg.pl/11/1e90ff/000000?text=+) **Fact:** *You can freely add your own config files too, all you need to do is to create a new file under `/config` and add your configuration to it. VELOX will know about this file and load it in the application. You can access your config via `Config::get('filename.whateverKeyYouWrote')`.*


---


## Classes

VELOX classes are divided in four namespaces:

* [`MAKS\Velox`](./classes)
* [`MAKS\Velox\Backend`](./classes/Backend)
* [`MAKS\Velox\Frontend`](./classes/Frontend)
* [`MAKS\Velox\Helper`](./classes/Helper)

The following table lists all available classes with their description:

| Class | Description |
| --- | --- |
| [`Velox\App`](./classes/App.php) | A class that serves as a basic service-container for VELOX. |
| <hr> | <hr> |
| [`Backend\Event`](./classes/Backend/Event.php) | A class that offers simple events handling functionality (dispatching and listening). |
| [`Backend\Config`](./classes/Backend/Config.php) | A class that loads everything from the `/config` directory and make it as an array that is accessible via dot-notation. |
| [`Backend\Router`](./classes/Backend/Router.php) | A class that serves as a router and an entry point for the application. |
| [`Backend\Globals`](./classes/Backend/Globals.php) | A class that serves as an abstraction/wrapper to work with superglobals. |
| [`Backend\Session`](./classes/Backend/Session.php) | A class that offers a simple interface to work with sessions. |
| [`Backend\Controller`](./classes/Backend/Controller.php) | An abstract class that serves as a base Controller that can be extended to make handlers for the router. |
| [`Backend\Database`](./classes/Backend/Database.php) | A class that represents the database and handles database operations. |
| [`Backend\Model`](./classes/Backend/Model.php) | An abstract class that serves as a base model that can be extended to create custom models. |
| [`Backend\Auth`](./classes/Backend/Auth.php) | A class that serves as an authentication system for users. |
| <hr> | <hr> |
| [`Frontend\Data`](./classes/Frontend/Data.php) | A class that serves as an abstracted data bag/store that is accessible via dot-notation. |
| [`Frontend\View`](./classes/Frontend/View.php) | A class that renders view files (Layouts, Pages, and Partials) with the ability to include additional files, divide page content into sections and cache rendered views. |
| [`Frontend\HTML`](./classes/Frontend/HTML.php) | A class that serves as a fluent interface to write HTML in PHP. It also helps with creating HTML elements on the fly. |
| [`Frontend\Path`](./classes/Frontend/Path.php) | A class that serves as a path resolver for different paths/URLs of the app. |
| <hr> | <hr> |
| [`Helper\Dumper`](./classes/Helper/Dumper.php) | A class that dumps variables and exception in a nice formatting. |
| [`Helper\Misc`](./classes/Helper/Misc.php) | A class that serves as a holder for various miscellaneous utility function. |


![RED](https://fakeimg.pl/11/f03c15/000000?text=+) **Note:** *This all what the VELOX package provides out of the box.*

![BLUE](https://fakeimg.pl/11/1e90ff/000000?text=+) **Fact:** *The `App`, `Event`, `Config`, `Router`, `Globals`, `Session`, `Database`, `Auth`, `Data`, `View`, `HTML`, `Path` classes are aliased on the root namespace for ease-of-use.*


---


## Functions

VELOX functions are divided into these files:

* [`helpers.php`](./functions/helpers.php): This is where helper functions for VELOX classes live, these are mainly functions that return an instance of class or alias some method on it.
* [`html.php`](./functions/html.php): This is where HTML helper functions live, these are nothing other than aliases for the most used PHP functions with HTML.

The following table lists all available functions and to which class/group they belong:

| Class/Group | Function(s) |
| --- | --- |
| `App::class` | `app()`, <br>`abort()`, <br>`terminate()` |
| `Event::class` | `event()` |
| `Config::class` | `config()` |
| `Router::class` | `router()`, <br>`handle()`, <br>`redirect()`, <br>`forward()` |
| `Database::class` | `database()` |
| `Globals::class` | `globals()` |
| `Session::class` | `session()`, <br>`flash()`, <br>`csrf()` |
| `Auth::class` | `auth()` |
| `View::class` | `view()`, <br>`render()`, <br>`render_layout()`, <br>`render_page()`, <br>`render_partial()`, <br>`section_push()`, <br>`section_reset()`, <br>`section_start()`, <br>`section_end()`, <br>`section_yield()`, <br>`include_file()` |
| `Data::class` | `data()`, <br>`data_has()`, <br>`data_get()`, <br>`data_set()` |
| `HTML::class` | `html()` |
| `Path::class` | `path()`, <br>`app_path_current()`, <br>`app_url_current()`, <br>`app_path()`, <br>`app_url()`, <br>`theme_path()`, <br>`theme_url()`, <br>`assets_path()`, <br>`assets_url()` |
| `Dumper::class` | `dd()`, <br>`dump()`, <br>`dump_exception()` |
| HTML Helpers | `he()`, `hd()`, `hse()`, `hsd()`, `st()`, `nb()` |


![BLUE](https://fakeimg.pl/11/1e90ff/000000?text=+) **Fact:** *You can freely add your own functions too, all you need to do is to create a new file under `/functions` and add your functions to it. VELOX will know about this file and load it in the application.*


---


## Commands

VELOX comes with some handy commands that you can use to do some repetitive tasks. You can execute these commands using the `php bin/{command-name}`.

The following table lists all available commands with their description.

| Command | Description |
| --- | --- |
| [`app-serve`](./bin/app-serve) | This command starts a development server. |
| [`app-mirror`](./bin/app-mirror) | This command mirrors the application in the `/public` directory. |
| [`config-cache`](./bin/config-cache) | This command caches the current configuration. |
| [`config-dump`](./bin/config-dump) | This command dumps the current configuration with syntax highlighting. |
| [`cache-clear`](./bin/cache-clear) | This command clears caches. |


CHANGELOG.md
# Changelog

All notable changes to **VELOX** will be documented in this file.

<br />

## [[1.5.8] - 2022-08-19](https://github.com/MarwanAlsoltany/velox/compare/v1.5.7...v1.5.8)

- Update `App` class:
    - Make constructor final.
    - Update `instance()` method.
    - Update `terminate()` method.
    - Update `shutdown()` method.
- Update `Globals` class:
    - Update `initialize()` method.
- Update `Database` class:
    - Add `#[\ReturnTypeWillChange]` attribute where necessary.
- Update `Element` class:
    - Add `#[\ReturnTypeWillChange]` attribute where necessary.
- Update `Dumper` class:
    - Update `VAR_DUMP_CONVERSIONS` class constant.
    - Update `dump()` method.
    - Update `varExport()` method.
    - Update `varDump()` method.
- Update `autoload.php`:
    - Add `$_VELOX` superglobal.
- PHP `8.1` support:
- Update tests:
    - Update `DumperTest` class.

<br />

## [[1.5.7] - 2022-06-12](https://github.com/MarwanAlsoltany/velox/compare/v1.5.6...v1.5.7)

- Update `Dumper` class:
    - Update `VAR_EXPORT_CONVERSIONS` class constant to fix an issue with multiline strings.
    - Refactor `exportExpression()` method and fix an issue when `var_export()` fails.

<br />

## [[1.5.6] - 2022-06-12](https://github.com/MarwanAlsoltany/velox/compare/v1.5.5...v1.5.6)

- Update `Exception` class:
    - Update `create()` method to fix an issue with returned namespace.
    - Update `throw()` method to accept `$previous` as `\Throwable` instead of `\Exception`.
    - Update `handle()` method.
- Update `Misc` class:
    - Update `transform()` method to add new cases.
- Update `Event` class:
    - Update `listen()` method to add the `$priority` argument.
    - Refactor `dispatch()` method to allow listeners to attach events when executed.
- Update `Dumper` class:
    - Add `VAR_EXPORT_CONVERSIONS` class constant.
    - Add `VAR_DUMP_CONVERSIONS` class constant.
    - Update `$styles` static property.
    - Add `varExport()` method.
    - Add `verDump()` method.
    - Refactor `exportExpression()` method and change visibility to `public`.
    - Refactor `getDumpingBlocks()` method.
    - Update `dump()` method.
- Update tests:
    - Add `EventTest` class.
    - Add `MiscTest` class.
    - Add `DumperTest` class.
- Update `composer.json`:
    - Update required `php` version.
    - Add required PHP extensions `ext-mbstring`.

<br />

## [[1.5.5] - 2021-12-28](https://github.com/MarwanAlsoltany/velox/compare/v1.5.4...v1.5.5)

- Add `Exception` class.
- Update the following classes to improve thrown exceptions classes and messages:
    - Update `App` class.
    - Update `Auth` class.
    - Update `Controller` class.
    - Update `Globals` class.
    - Update `Router` class.
    - Update `Database` class.
    - Update `Model` class.
    - Update `DBAL` class.
    - Update `Element` class.
- Update `Dumper` class:
    - Update `$styles` property to improve exceptions page styling.
    - Update `exportExpression()` method.
    - Update `exportExpressionWithSyntaxHighlighting()` method.
    - Update `dump()` method to fix an issue with block styling.
    - Update `dumpException()` method to add code snippet for each item in the stacktrace.
    - Add `highlightedFile()` method.
    - Add `tabulatedStacktrace()` method.
- Update `Config` class:
    - Add `getReference()` method.
    - Update `getAll()` method.
- Update `Data` class:
    - Update `load()` method to make use of `Config::getReference()` instead of resetting the config array.
- Update `loader.php`:
    - Update error handler function.
    - Update exception handler function to allow for rendering specific error pages using exception code.
- Update tests:
    - Add `ExceptionTest` class.

<br />

## [[1.5.4] - 2021-12-20](https://github.com/MarwanAlsoltany/velox/compare/v1.5.3...v1.5.4)

- Update `App` class:
    - Remove `__get()` magic method.
- Update `Database` class:
    - Update `mock()` method.
- Update `View` class:
    - Refactor class by extracting parts of it to `View/Compiler`.
- Add `Compiler` class.
- Update `Session` class:
    - Refactor class by extracting parts of it to `Session/Flash` and `Session/CSRF`.
- Add `Session\Flash` class.
- Add `Session\CSRF` class.
- Update tests:
    - Update `AppTest` class.
    - Update `EngineTest` class.

<br />

## [[1.5.3] - 2021-12-15](https://github.com/MarwanAlsoltany/velox/compare/v1.5.2...v1.5.3)

- Update `Engine` class:
    - Update `getCompiledContent()` method.
    - Update `printVariables()` method to print escaped and unescaped variables.
    - Remove `printUnescapedVariables()` method.
    - Refactor `wrapControlStructures()` method.
- Update `Router` class:
    - Update `start()` method to allow for `ON_START` event to be dispatched when CSRF check fails.
    - Rename `doEchoResponse()` method to `respond()`.

<br />

## [[1.5.2] - 2021-12-14](https://github.com/MarwanAlsoltany/velox/compare/v1.5.1...v1.5.2)

- Update `App` class:
    - Update `abort()` method.
- Update `Dumper` class:
    - Add `$styles` property.
    - Add `getDumpingBlocks()` method.
    - Refactor `dump()` method.
    - Update `dumpException()` method.
    - Update `exportExpressionWithSyntaxHighlighting()` method.
- Update `Router` class:
    - Add `HANDLER_ROUTE` class constant.
    - Add `MIDDLEWARE_ROUTE` class constant.
    - Change visibility of `getValidParameters()`, `getRoutePath()`, `getRouteRegex()`, `getRouteArguments()`, `doEchoResponse(`) methods from private to protected.
- Update `Element` class:
    - Rename class from `Base` to `Element`. This shouldn't introduce any problems as the `Base` class was marked as `@internal`.
    - Fix `set()` method return type hint.
    - Fix `setAttributes()` method return type hint.

<br />

## [[1.5.1] - 2021-12-12](https://github.com/MarwanAlsoltany/velox/compare/v1.5.0...v1.5.1)

- Update `App` class:
    - Add `$instance` property.
    - Update `instance()` method.
- Update `Engine` class:

classes/App.php
<?php

/**
 * @author Marwan Al-Soltany <MarwanAlsoltany@gmail.com>
 * @copyright Marwan Al-Soltany 2021
 * For the full copyright and license information, please view
 * the LICENSE file that was distributed with this source code.
 */

declare(strict_types=1);

namespace MAKS\Velox;

use MAKS\Velox\Backend\Exception;
use MAKS\Velox\Backend\Event;
use MAKS\Velox\Backend\Config;
use MAKS\Velox\Backend\Router;
use MAKS\Velox\Backend\Globals;
use MAKS\Velox\Backend\Session;
use MAKS\Velox\Backend\Database;
use MAKS\Velox\Backend\Auth;
use MAKS\Velox\Frontend\Data;
use MAKS\Velox\Frontend\View;
use MAKS\Velox\Frontend\HTML;
use MAKS\Velox\Frontend\Path;
use MAKS\Velox\Helper\Dumper;
use MAKS\Velox\Helper\Misc;

/**
 * A class that serves as a basic service-container for VELOX.
 * This class has most VELOX classes as public properties:
 * - `$auth`: Instance of the `Auth` class.
 * - `$event`: Instance of the `Event` class.
 * - `$config`: Instance of the `Config` class.
 * - `$router`: Instance of the `Router` class.
 * - `$globals`: Instance of the `Globals` class.
 * - `$session`: Instance of the `Session` class.
 * - `$database`: Instance of the `Database` class.
 * - `$data`: Instance of the `Data` class.
 * - `$view`: Instance of the `View` class.
 * - `$html`: Instance of the `HTML` class.
 * - `$path`: Instance of the `Path` class.
 * - `$dumper`: Instance of the `Dumper` class.
 * - `$misc`: Instance of the `Misc` class.
 *
 * Example:
 * ```
 * // create an instance
 * $app = new App();
 * // get an instance of the `Config` class via public property access notation
 * $app->config;
 * // or via calling a method with the same name
 * $app->config()->get('global');
 * ```
 *
 * @package Velox
 * @since 1.0.0
 * @api
 *
 * @method static void handleException(\Throwable $expression) This method is available only at shutdown.
 * @method static void handleError(int $code, string $message, string $file, int $line) This method is available only at shutdown.
 */
class App
{
    /**
     * This event will be dispatched on app termination. Note that this event can be dispatched multiple times in app life-cycle.
     * This event will not be passed any arguments.
     *
     * @var string
     */
    public const ON_TERMINATE = 'app.on.terminate';

    /**
     * This event will be dispatched on app shutdown. Note that this event is dispatched only once in app life-cycle.
     * This event will not be passed any arguments.
     *
     * @var string
     */
    public const ON_SHUTDOWN = 'app.on.shutdown';


    /**
     * The class singleton instance.
     */
    protected static self $instance;


    public Event $event;

    public Config $config;

    public Router $router;

    public Globals $globals;

    public Session $session;

    public Database $database;

    public Auth $auth;

    public Data $data;

    public View $view;

    public HTML $html;

    public Path $path;

    public Dumper $dumper;

    public Misc $misc;

    protected array $methods;

    protected static array $staticMethods;


    /**
     * Class constructor.
     */
    final public function __construct()
    {
        empty(static::$instance) && static::$instance = $this;

        $this->event    = new Event();
        $this->config   = new Config();
        $this->router   = new Router();
        $this->globals  = new Globals();
        $this->session  = new Session();
        $this->database = Database::instance();
        $this->auth     = Auth::instance();
        $this->data     = new Data();
        $this->view     = new View();
        $this->html     = new HTML();
        $this->path     = new Path();
        $this->dumper   = new Dumper();
        $this->misc     = new Misc();

        $this->methods  = [];
        static::$staticMethods = [];
    }

    public function __call(string $method, array $arguments)
    {
        $class = static::class;

        try {
            return isset($this->methods[$method]) ? $this->methods[$method](...$arguments) : $this->{$method};
        } catch (\Exception $error) {
            Exception::throw(
                'UndefinedMethodException:BadMethodCallException',
                "Call to undefined method {$class}::{$method}()",
                $error->getCode(),
                $error,
            );
        }
    }

    public static function __callStatic(string $method, array $arguments)
    {
        $class = static::class;

        if (!isset(static::$staticMethods[$method])) {
            Exception::throw(
                'UndefinedStaticMethodException:BadMethodCallException',
                "Call to undefined static method {$class}::{$method}()"
            );
        }

        return static::$staticMethods[$method](...$arguments);
    }


    /**
     * Returns the singleton instance of the `App` class.
     *
     * NOTE: This method returns only the first instance of the class
     * which is normally the one that was created during application bootstrap.
     *
     * @return static
     *
     * @since 1.4.0
     */
    final public static function instance(): self
    {
        empty(static::$instance) && static::$instance = new static();

        return static::$instance;
    }

    /**
     * Extends the class using the passed callback.
     *
     * @param string $name Method name.
     * @param callable $callback The callback to use as method body.
     *
     * @return callable The created bound closure.
     */
    public function extend(string $name, callable $callback): callable
    {
        $method = \Closure::fromCallable($callback);
        $method = \Closure::bind($method, $this, $this);

        return $this->methods[$name] = $method;
    }

    /**
     * Extends the class using the passed callback.
     *
     * @param string $name Method name.
     * @param callable $callback The callback to use as method body.
     *
     * @return callable The created closure.
     */
    public static function extendStatic(string $name, callable $callback): callable
    {
        $method = \Closure::fromCallable($callback);
        $method = \Closure::bind($method, null, static::class);

        return static::$staticMethods[$name] = $method;
    }

    /**
     * Logs a message to a file and generates it if it does not exist.
     *
     * @param string $message The message wished to be logged.
     * @param array|null $context An associative array of values where array key = {key} in the message (context).
     * @param string|null $filename [optional] The name wished to be given to the file. If not provided `{global.logging.defaultFilename}` will be used instead.
     * @param string|null $directory [optional] The directory where the log file should be written. If not provided `{global.logging.defaultDirectory}` will be used instead.
     *
     * @return bool True on success (if the message was written).
     */
    public static function log(string $message, ?array $context = [], ?string $filename = null, ?string $directory = null): bool
    {
        if (!Config::get('global.logging.enabled', true)) {
            return true;
        }

        $hasPassed = false;

        if (!$filename) {
            $filename = Config::get('global.logging.defaultFilename', sprintf('autogenerated-%s', date('Ymd')));
        }

        if (!$directory) {
            $directory = Config::get('global.logging.defaultDirectory', BASE_PATH);
        }

        $file = Path::normalize($directory, $filename, '.log');

        if (!file_exists($directory)) {
            mkdir($directory, 0744, true);
        }

        // create log file if it does not exist
        if (!is_file($file) && is_writable($directory)) {
            $signature = 'Created by ' . __METHOD__ . date('() \o\\n l jS \of F Y h:i:s A (Ymdhis)') . PHP_EOL . PHP_EOL;
            file_put_contents($file, $signature, 0);
            chmod($file, 0775);
        }

        // write in the log file
        if (is_writable($file)) {
            clearstatcache(true, $file);
            // empty the file if it exceeds the configured file size
            $maxFileSize = Config::get('global.logging.maxFileSize', 6.4e+7);
            if (filesize($file) > $maxFileSize) {
                $stream = fopen($file, 'r');
                if (is_resource($stream)) {
                    $signature = fgets($stream) . 'For exceeding the configured {global.logging.maxFileSize}, it was overwritten on ' . date('l jS \of F Y h:i:s A (Ymdhis)') . PHP_EOL . PHP_EOL;
                    fclose($stream);
                    file_put_contents($file, $signature, 0);
                    chmod($file, 0775);
                }
            }

            $timestamp = (new \DateTime())->format(DATE_ISO8601);
            $message   = Misc::interpolate($message, $context ?? []);

            $log = "$timestamp\t$message\n";

            $stream = fopen($file, 'a+');
            if (is_resource($stream)) {
                fwrite($stream, $log);
                fclose($stream);
                $hasPassed = true;
            }
        }

        return $hasPassed;
    }

    /**
     * Aborts the current request and sends a response with the specified HTTP status code, title, and message.
     * An HTML page will be rendered with the specified title and message.
     * If a view file for the error page is set using `{global.errorPages.CODE}`,
     * it will be rendered instead of the normal page and passed the `$code`, `$title`, and `$message` variables.
     * The title for the most common HTTP status codes (`200`, `401`, `403`, `404`, `405`, `500`, `503`) is already configured.
     *
     * @param int $code The HTTP status code.
     * @param string|null $title [optional] The title of the HTML page.
     * @param string|null $message [optional] The message of the HTML page.
     *
     * @return void
     *
     * @since 1.2.5
     */
    public static function abort(int $code, ?string $title = null, ?string $message = null): void
    {
        $http = [
            200 => 'OK',
            401 => 'Unauthorized',
            403 => 'Forbidden',
            404 => 'Not Found',
            405 => 'Not Allowed',
            500 => 'Internal Server Error',
            503 => 'Service Unavailable',
        ];

        $title    = htmlspecialchars($title ?? $code . ' ' . $http[$code] ?? '', ENT_QUOTES, 'UTF-8');
        $message  = htmlspecialchars($message ?? '', ENT_QUOTES, 'UTF-8');

        try {
            $html = View::render((string)Config::get("global.errorPages.{$code}"), compact('code', 'title', 'message'));
        } catch (\Throwable $e) {
            $html = (new HTML(false))
                ->node('<!DOCTYPE html>')
                ->open('html', ['lang' => 'en'])
                    ->open('head')
                        ->title((string)$code)
                        ->link(null, [
                            'href' => 'https://cdn.jsdelivr.net/npm/bulma@latest/css/bulma.min.css',
                            'rel' => 'stylesheet'
                        ])
                    ->close()
                    ->open('body')
                        ->open('section', ['class' => 'section is-large has-text-centered'])
                            ->hr(null)
                            ->h1($title, ['class' => 'title is-1 is-spaced has-text-danger'])
                            ->condition(strlen($message))
                            ->h4($message, ['class' => 'subtitle'])
                            ->hr(null)
                            ->a('Reload', ['class' => 'button is-warning is-light', 'href' => 'javascript:location.reload();'])
                            ->entity('nbsp')
                            ->entity('nbsp')
                            ->a('Home', ['class' => 'button is-success is-light', 'href' => '/'])
                            ->hr(null)
                        ->close()
                    ->close()
                ->close()
            ->return();
        } finally {
            http_response_code($code);
            echo $html;

            static::terminate();
        }
    }


    /**
     * Terminates (exits) the PHP script.
     * This function is used instead of PHP `exit` to allow for testing `exit` without breaking the unit tests.
     *
     * @param int|string|null $status The exit status code/message.
     * @param bool $noShutdown Whether to not execute the shutdown function or not.
     *
     * @return void This function never returns. It will terminate the script.
     * @throws \Exception If `EXIT_EXCEPTION` is defined and truthy.
     *
     * @since 1.2.5
     */
    public static function terminate($status = null, bool $noShutdown = true): void
    {
        Event::dispatch(self::ON_TERMINATE);

        if (defined('EXIT_EXCEPTION') && EXIT_EXCEPTION) {
            throw new \Exception(empty($status) ? 'Exit' : 'Exit: ' . $status);
        }

        // @codeCoverageIgnoreStart
        if ($noShutdown) {
            // app shutdown function checks for this variable
            // to determine if it should exit, see bootstrap/loader.php
            Misc::setArrayValueByKey($GLOBALS['_VELOX'], 'TERMINATE', true);
        }

        exit($status);
        // @codeCoverageIgnoreEnd
    }

    /**
     * Shuts the app down by terminating it and executing shutdown function(s).
     * The triggered shutdown functions can be normal shutdown functions registered,
     * using `register_shutdown_function()` or the `self::ON_SHUTDOWN` event.
     *
     * @return void
     *
     * @internal This method is to be used by the framework and not the user.
     * @since 1.4.2
     *
     * @codeCoverageIgnore
     */
    public static function shutdown(): void
    {
        Event::dispatch(self::ON_SHUTDOWN);

        Misc::setArrayValueByKey($GLOBALS['_VELOX'], 'SHUTDOWN', false);

        exit;
    }
}
